### PR TITLE
Quick setup of Priority roads

### LINKED ISSUE #541: shortcut to fix traffic rules where minor road joins a main avenue
in real life and also in simcity when a small road joins a main avenue, the separation barrier in the middle remains unbroken and people can only turn right (the only advantage of simcity over cities skylines!!!). we already have an asset that visually conserves the junction barrier in such situations

previewfile_1319965985

But TMPE is required to make it function as it should. It can be a bit tedious to apply all the traffic rules particularly when fixing cities where many minor roads are joining one avenue. 

so i suggest a shortcut: choose junction priority tool then press **ctrl+click** on a junction to set all rules regarding the above scenario:
 - remove zebra crossings on the avenue
 - cars on the avenue can enter block junctions
 - no left turns
 - give way entering the avenue

Let me clarify I prefer this feature apply to all big roads such as this:
median rules

### PR DESCRIPTION
**Pull request by kianzarrin**
Fixes #541
related #542
related #568: undo delete option.
Usage:
- select priority sign tool.
- Ctrl+click to quick setup priority junction
- Ctrl+Shift+Click to do the same to multiple segments.

Situations handled: *
- minor road joins main avenue 
- Split 2 way road into one-way roads 
- semi-roundabout 
- no yield required when lane arithmetic is observed
 
 \* see https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/541#issuecomment-557296554 for test city and more information. 

Crowdin: 
~~I asked krzychu to add crowding keys. I also modified tutorial and hotkey corwdin texts (waiting for approval)
FYI:~~
EDIT: @krzychu124 did not respond. Can anyone please add these keys?
```
Options.csv:
"MassEdit.Group.Priority roads"->"Priority roads"
"Priority roads.Option:Allow far turns"->"Allow cars to take far turn from/into main road (not recommended)"
"Priority roads.Tooltip:Allow far turns"->"If enabled, lane arrows are untouched and cars can cross road median (does not apply when splitting avenue)."
"Priority roads.Option:Enter blocked yield road"->"Allow cars on yield road to enter blocked main road"
"Priority roads.Option:Stop signs on entry"->"Use stop signs when entering main road"
"Priority roads.Tooltip:Stop signs on entry"->"If enabled, puts stop signs on roads entering the main road. If disabled, puts yield signs instead"
"Priority roads.Option:Allow pedestrian crossings on main road"->"Allow pedestrian crossings on main road"
```

I made the mass edit overly to last a second after processing is done so that the user has some visual feedback that something has happened.

### CODE REVIEW

**Comment by kianzarrin** [comment]
I thought about providing these options but i filtered it down to the ones I deemed useful. What do you think:
```
"X" means I do not want it

rabout_DedicatedExitLanes
rabout_SwitchLanesYeildR X
rabout_SwitchLanesMainR X
rabout_StayInLaneMainR
rabout_StayInLaneNearRabout
rabout_NoCrossMainR
rabout_NoCrossYeildR
rabout_PrioritySigns

avn_NoCrossMainR
avn_NoCrossYield X
avn_NoLeftTurns
avn_EnterBlockedMain X
avn_EnterBlockedYeild
avn_StopEntry
avn_PrioritySigns X
avn_SwitchLanesMain X
avn_SwtichLanesYeild X

```

**Comment by aubergine10** [comment]
Do you have descriptions for each of those features you listed in https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/621#issuecomment-576553398

**Comment by aubergine10** [comment]
In tooltip, Ctrl+Shift+Click is mentioned, but not the Ctrl+Click

image

**Comment by aubergine10** [comment]
Pressing just Shift key highlights route (tooltip says Ctrl+Shift)

image

**Comment by aubergine10** [comment]
Holding Ctrl over junction doesn't change the visual feedback.

normal vs. ctrl

In clip above, I first hovered normally (no modifier key) then hovered with Ctrl pressed.

Maybe change colour if Ctrl pressed so user can see difference between Ctrl+hover and normal hover?

**Comment by aubergine10** [comment]
There's no indication of what will happen _before_ I click with either Shift or Ctrl pressed.

If the modifier keys are pressed, it would be nice if overlays could indicate what will happen (or change) when I click.

**Comment by aubergine10** [comment]
After Ctrl+Click this junction the overlays show but it's difficult to work out which roads they apply to (particularly the junction restriction overlays).

screenshot

**Comment by aubergine10** [comment]
IMO disabling ped crossings should require an additional modifier key (Alt?) otherwise users could drastically restrict crossing points for pedestrians/cyclists and that could have big impact on pathfinding (there's no guarantee that users will think to add bridges over roads).

Also, the temporary overlay timer seems to short; the overlays disappear before I get chance to have good look at them. Maybe increase to 5 seconds or, when changes applied to route, 15 seconds or more?

**Comment by aubergine10** [comment]
When applying to route, it looks like it's selecting the route based on direction however in some cases that is not idea.

Consider this road where the main road turns sharply but the rural road continues forwards:

image

When I Shift to select the route, this happens:

bandicam 2020-01-31 22-42-54-686

Maybe it should try and follow same type of road, checking first the "expected onwards" direction and if it finds different road first check if any left/right roads seem to be be continuation of original road?

> Maybe Alt modifier could affect that behaviour (Alt = try and stick to original road even if it goes left/right)?

**Comment by aubergine10** [comment]
Avoid negative option effects:

image

Would be better as:

* Pedestrians can cross to center of roundabout (off by default)
* Pedestrians can cross where roads enter the roundabout (on by default)

**Comment by aubergine10** [comment]
I Control+Click this road and it doesn't add any priority signs:

image

The road going left/right is same road type, so ideally should be seen as main road. Road entering from bottom of T-Junction also had a vanilla stop sign:

image

What I was expecting:

image

Note again terrible placement of vehicle restrictions persistent overlay. The 'enter blocked junction' is applied to one-way the road entering from the right.

**Comment by kianzarrin** [comment]
> @aubergine10  ick this junction the overlays show but it's difficult to work out which roads they apply to (particularly the junction restriction overlays).

If you go to options and activate the relevant overlays you will see that here too it's difficult to work out which roads they apply to 

If you go to junction restrictions but do not select any junction again you will see the overlays but here too it's difficult to work out which roads they apply to .

Only if you select a junction, the overlays will go to a good position.

Nothing to do with this issue, but if you like you can create an issue to address this problem.

**Comment by kianzarrin** [comment]
@aubergine10 
> Avoid negative option effects:
> Would be better as:
>     * Pedestrians can cross to center of roundabout (off by default)
>     * Pedestrians can cross roads entering/exiting the roundabout (on by default)

The default behavior is pedestrians can cross all junctions. I am trying to override the default behavior.
If the options are turned off, my code will not alter the previous state  pedestrian crossings. which is different than turning them on and off.

for example consider the scenario if the user has built  a bridge over one of the roads connected to the roundabout and therefore has banned pedestrian crossings on that one entry. but other entries do allow pedestrian crossings. What should the user do if he does not want my code to touch the crosswalks? 

It would be wired to have an option to say pedestrians can walk to the center of the roundabout! why would I enable crossings to the center if they were previously banned.

If you really do not like negatives I can modify it to something like this.
 - Fix pedestrian crossings to the center of the roundabout or
 - Ban pedestrian crossings to the center of roundabout.

In any case this is about roundabout changes from the other pull request

**Comment by kianzarrin** [comment]
> Note again terrible placement of vehicle restrictions persistent overlay. The 'enter blocked junction' is applied to one-way the road entering from the right.

hmmm ... this looks like to semi roundabout case in my code. I am surprised it did not work. I will look into this.

**Comment by kianzarrin** [comment]
@aubergine10 
> When applying to route, it looks like it's selecting the route based on direction however in some cases that is not ideal.
> What I was expecting as end user:
> 
> image
> 
> Maybe it should try and follow same type of road, checking first the "expected onwards" direction and if it finds different road first check if any left/right roads seem to be be continuation of original road?
> 
> Maybe Alt modifier could affect that behaviour (Alt = try and stick to original road type even if it goes left/right)?

I have some bad news and some good news.

### Bad news:
the previous shit-click also has this "problem". unlike my roundabout review I have not add clever traversal but use the already existing traversal function. whats more is that if the road turns I ignore lane arrows. Reasons:
 -  high priority main roads do not have sharp turns.(no normally anyway)
 - there are too many different possibilities. I just found it simpler to check for common cases where the road is straight.

adding alt modifier is too much. too many modifiers can confuse the user.

The issue you mention was on my mind when writing the code but I decided to ignore it because:

### Good news: 
when #631 is merged you can select your path. considering road selection is a thing I decided to keep my traversal code simple. 

### Bad news again:
But even when #631 is complete I still ignore the junction. I should not try to do everything in one review. This condition should be handled in another issue.

**Comment by aubergine10** [comment]
> hmmm ... this looks like the semi roundabout case in my code. I am surprised it did not work. I will look into this.

Could it be caused by the fact I had lane connectors on that junction prior to Ctrl+Click?

File: TLM/TLM/UI/SubTools/PrioritySignsTool.cs [hunk 1]
```diff
@@ -165,18 +160,84 @@ public PrioritySignsTool(TrafficManagerTool mainTool)
         public override void OnToolGUI(Event e) { }
 
         /// <summary>
-        /// show overlay for other subtools influced by mass edit.
+        /// Thread safe handling of mass edit overlay.
         /// </summary>
-        public static bool showMassEditOverlay = false;
+        public static class MassEditOVerlay {
```
[the remark this answers is not in the document]
Comment (krzychu124) [reply TLM/TLM/UI/SubTools/PrioritySignsTool.cs]: > this is all being handled in the same simulation thread.

No it's not! `OnPrimaryClickOverlay` like any other overlay related feature is called from main UI Update thread, not game simulation...

**Comment by kianzarrin** [comment]
@aubergine10 
> Could it be caused by the fact I had lane connectors on that junction prior to Ctrl+Click?

its a bug with LHT

RHT works:
Screenshot (390)

I am on it.

**Comment by kianzarrin** [comment]
@aubergine10 I fixed some intermittent bugs and merge issues. semi roundabout problem with LHT is also fixed. Should work now.

**Comment by aubergine10** [comment]
Looking better!

The temporary overlays (eg. junct restrictions) still disappear too fast; should be at least 3-5 seconds for the Ctrl+Click and maybe 10-15 seconds for Ctrl+Shift+Click.

----

The "left" in this option caption is going to be problematic for LHT cities, it's already seeped in to the planned translation:
image

> "Priority roads.Option:Allow left turns"->"Allow cars to turn left from/into main road (not recommended)"

Consider "far-side" or "cut across oncoming traffic" so it's direction agnostic?

----

Would there be any merit in having an option to define whether cims can cross main road when doing Ctrl+Shift+Click? For a single junction, I can happily prevent them crossing, but for entire route it's a bit risky.

Consider this single road, for example, which has about a dozen roads joining it... If I disable crossings all the way down the road, it's going to be a nightmare for cims to navigate across it:

image

Could maybe have (enabled by default):

> "Priority roads.Option:Allow pedestrian crossings on main road on long routes"->"Allow pedestrian crossings on main road on long routes"

**Comment by aubergine10** [comment]
> I fixed some intermittent bugs and merge issues. semi roundabout problem with LHT is also fixed. Should work now.

https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/621#issuecomment-580949890 confirmed as fixed.

**Comment by brunoais** [comment]
> > "Priority roads.Option:Allow pedestrian crossings on main road on long routes"->"Allow pedestrian crossings on main road on long routes"

I don't like that one... What about:
> "Allow pedestrian crossings on main road on routes longer than (default: XX)" 

and then being able to set a unit. "long routes" appears to be too subjective and it is fixed without possibility to change. Then, just give a sensible default.

**Comment by aubergine10** [comment]
> "Allow pedestrian crossings on main road on routes longer than (default: XX)"
> and then being able to set a unit. "long routes" appears to be too subjective and it a fixed distance without possibility to change. Then, just give a sensible default.

In-game, user generally has no sense of length or dimensions of anything. Stuff is just "big or small", not "x km". Without using something like Measure It! mod, user wouldn't have a clue what length to set.

The issue I have with how it works currently is that the user could unknowingly cause mass problems due to disabled ped crossings, and they might not even be aware that some junction far off-screen had it's crossings disabled. In the screenshot I posted, I'd have to go manually check every junction along that massive road to work out if there were going to be issues with ped crossings.

Note that all the other settings (priority signs, and the other junction restrictions) won't break pathfinding. So it's specifically the ped crossings that are the issue. Maybe the solution is to make it so the ped crossings can only be disabled when applying to single junction? That way there's no chance of user making massive long road uncrossable.

**Comment by kianzarrin** [comment]
@aubergine10 
> 
> Would there be any merit in having an option to define whether cims can cross main road when doing Ctrl+Shift+Click? For a single junction, I can happily prevent them crossing, but for entire route it's a bit risky.
> 
> Consider this single road, for example, which has about a dozen roads joining it... If I disable crossings all the way down the road, it's going to be a nightmare for cims to navigate across it:
> 
> image
> 
> Could maybe have (enabled by default):
> 
> > "Priority roads.Option:Allow pedestrian crossings on main road on long routes"->"Allow pedestrian crossings on main road on long routes"

Just imagine! you have a junction so important that cars may not take the far turn but pedestrians can walk right across! I have never seen a junction like this in my life. High priority roads have pedestrian bridges or traffic lights. In your case it would be more appropriate to use shift-click to apply only priority signs. 

It just occurred to me we can make the "priority road" VS "high priority road" terminology a thing.

**Comment by brunoais** [comment]
> It just occurred to me we can make the "priority road" VS "high priority road" terminology a thing.
That may be the solution we need

**Comment by aubergine10** [comment]
I love the conepot of priority vs. high priority.

In UK, pedestrians always have right of way (humans >> machines), except where it is simply too dangerous for them to be crossing roads (safety > painful death). Examples:

* "B" road: Pedestrians cross wherever they want
* "A" road: Pedestrians can cross wherever they want, but they usually head for marked crossing point ([zebra](https://en.wikipedia.org/wiki/Zebra_crossing) or [pelican](https://en.wikipedia.org/wiki/Pelican_crossing) crossing) if there is one nearby
* "Dual carriageway": (Car) Priority road - pedestrians should use marked crossings or ped bridges/tunnels
* "Motorway": High (car) priority roads - pedestrians must use bridges/tunnels

We could use `Alt` modifier to distinguish between high and normal car priority:

* `Control+Click` = normal car priority
* `Alt+Control+Click` = high car priority

And add `Shift` to those shortcuts for "apply to route" mode.

**Comment by aubergine10** [comment]
^ Updated comment above several times might need re-reading

**Comment by aubergine10** [comment]
For docs, could you elaborate on what this means:

> no yield required when lane arithmetic is observed

**Comment by kianzarrin** [comment]
@aubergine10 
> EDIT: Just realised there is no temporary display of overlays; it was just be holding and releasing Ctrl key lol

I don't know the context you are taking about but the mass edit overlay should show for a short time after you CTRL-SHIFT-CLICK

**Comment by aubergine10** [comment]
> I don't know the context you are taking about but the mass edit overlay should show for a short time after you CTRL-SHIFT-CLICK

I was referring to just the single-junction `Ctrl+Click` thinking that had delay on the overlays (it was actually just delay between me ctrl+click and me releasing ctrl key).

Will retest the ctrl+shift+click in a moment.

**Comment by kianzarrin** [comment]
> I was referring to just the single-junction `Ctrl+Click` thinking that had delay on the overlays (it was actually just delay between me ctrl+click and me releasing ctrl key).

`Ctrl+Click` as well as `CTRL+Shift+Click` should continue to show the overlay a few seconds after the click (after all the processing is done to be more accurate).

CTRL+Shift+Click is like pressing CTRL+Click on every junction on the highlighted road.

**Comment by aubergine10** [comment]
Just tested again:

* Ctrl+Click = about 1.5-2 seconds
* Ctrl+Shift+Click = same

Maybe **Ctrl+Click** should be `3` seconds, and **Ctrl+Shift+Click** should be `3 * num_junctions` up to a max of `15` seconds?

**Comment by kianzarrin** [comment]
> Just tested again:
>     * Ctrl+Click = about 1.5-2 seconds
>     * Ctrl+Shift+Click = same

Its one second after all processing is done:
```
            // keep active for one more second so that the user
            // has a chance to see the new traffic rules.
            MassEditOVerlay.SetTimer(1);
```
> Maybe **Ctrl+Click** should be `3` seconds, and **Ctrl+Shift+Click** should be `3 * num_junctions` up to a max of `15` seconds?

why? do you know its one second AFTER all processing is done and not before

**Comment by aubergine10** [comment]
Ok, we can revisit the timer thing at later date after overlays are optimised.

What about this: https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/621#issuecomment-582545064

**Comment by kvakvs** [comment]
Just a side note
In real world road priority for pedestrians is tightly coupled with speed limit on that street and how many lanes that road has. So say under 20mph/30kmph that will be pedestrians priority; Then 30mph/50kmph it is ok to cross (always can cross if there is a traffic light), and over that pedestrians can't cross at all without a traffic light being green for them.
How to display that is another question :)

**Comment by kianzarrin** [comment]
@aubergine10 I am really bad at road numbers. It seems British numbering system is much better than Israel:
https://en.wikipedia.org/wiki/Great_Britain_road_numbering_scheme

It also depends on number of digits the A road has.

Examples of A road and B road on google:

- https://en.wikipedia.org/wiki/A4226_road#/media/File:A4226_near_Weycock_Cross.jpg
- https://upload.wikimedia.org/wikipedia/commons/0/0a/A4226_near_Weycock_Cross.jpg
-  https://www.google.com/maps/@52.421688,-1.7863173,3a,75y,227.95h,92.1t/data=!3m6!1e1!3m4!1s5dJsrtpJ0MVisNblDTPVxQ!2e0!7i13312!8i6656
- https://goo.gl/maps/huBkVSuZ9FdS4RWZ6

In the roads above cars can mostly turn left and right: 
in this one left turn is forbidden but the road joining it is auxiliary roadSo you should not use the high priority modifier (CTRL). 

In my road selection panel I can provide profiles(maybe even customizable) for multiple different levels of roads.

but for the sake of this review I think what I am doing is good enough. you can always change the options.

**Comment by aubergine10** [comment]
> but we already have shift click for low priority (priority sings only) CTRL for high priority.

True, but it doesn't set the **enter blocked junction** stuff. 

Also, `Ctrl` seems to be setting it's place as the "multitool tool". Whereas `Shift` will just propagate basic tool down the road, `Ctrl+Shift` would apply other tools too (junction restricitons, lane arrows, etc...) hence desire for `Alt+Ctrl` and `Alt+Ctrl+Shift`.

* `Shift` = single tool propagate
* `Ctrl+Shift` = multi-tool propagate
* `Alt+Ctrl+Shift` = alternate multi-tool propagate

**Comment by kianzarrin** [comment]
> Also, `Ctrl` seems to be setting it's place as the "multitool tool". Whereas `Shift` will just propagate basic tool down the road, `Ctrl+Shift` would apply other tools too (junction restricitons, lane arrows, etc...) hence desire for `Alt+Ctrl` and `Alt+Ctrl+Shift`.

in comparison CTRL+SHIFT does the following:
- remove two crossings from main road
- add yield sign on connected road[s]
- ban far turn on connected road[s]
- ban far turn[s] on main road
- go through blocked junction on main road.
Even remembering all this requires a lot of energy. let alone applying them all to several junctions with no mistakes.

so ALT+CTRL+Shift only sets priority signs and go through blocked junction? that is something that can be easily done manually. if the objective is to set go through blocked junction on multiple junctions at once it should be part of bulk edit feature in junction restriction.

If you still feel strongly about ALT+CTRL+Shift I can add it but in my opinion its too much and it only confuses people.

**Comment by aubergine10** [comment]
With the way it currently stands, if I want alternate crossing mode (either remove or don't remove, ie. opposite of whatever is set in mod options) I have two options which are both clunky:

1. Change mod options:
    * Exit all the way to pause menu
    * Select options, find TMPE in list
    * Go to policies tab, change option
    * Exit from options
    * Go back in to tool, use the shortcut
    * or...
2. Switch tools:
    * Apply the usual shortcut in priority tool
    * Switch to junction restrictions tool
    * Select junction
    * Change crossings
    * Potentially for several junctions

Having `Alt` key temporarily toggle what happens to crossings over main road would be much easier.

From user perspective, I don't think they will try and remember what `Ctrl` modifier does (as it's a bit differnet depending on each junction). The `Alt` modifier will just be: "Whatever `Ctrl` would do, but invert what happens to crossings over main road"

Like, if I want normal behaviour (remove crossings) along route, except on couple of junctions, I could:

* `Ctrl+Shift+Click` the route
* `Alt+Ctrl+Click` the two junctions

No need to change tools or mod options, just very fluid and quick interaction with my roads using one tool.

**Comment by aubergine10** [comment]
BTW, I'm already using this bulk applicator all over my city, it's awesome!!

**Comment by kianzarrin** [comment]
> @aubergine10  Having Alt key temporarily toggle what happens to crossings over main road would be much easier.
From user perspective, I don't think they will try and remember what Ctrl modifier does (as it's a bit differnet depending on each junction). The Alt modifier will just be: "Whatever Ctrl would do, but invert what happens to crossings over main road"
Like, if I want normal behaviour (remove crossings) along route, except on couple of junctions, I could:
    Ctrl+Shift+Click the route
    Alt+Ctrl+Click the two junctions

What happens to lane arrows?

having the alt key doing opposite of what crossings do in the options is strange. maybe remove that option?  maybe provide a panel like the one in roundabout mod? maybe we can do this in the road selection panel? I feel like we are rushing in the ALT modifier. lets open an issue about this so that we can have some proper discussion. It would be wrong to shove it in a code review last minute.

**Comment by brunoais** [comment]
> EDIT: isn't it illegal to block junctions when cars joining the road can take the far turn?

Where I live, that's only illegal if there's markings on the ground meaning as such. Otherwise, it's legal

**Comment by kianzarrin** [comment]
Translation keys work.
Screenshot (414)

**Comment by kianzarrin** [comment]
@aubergine10 I created a branch with the ALT key just for you! https://github.com/kianzarrin/Cities-Skylines-Traffic-Manager-President-Edition/tree/AltPriorityRoad

the alt key force applies pedestrian crossing regardless of the options.
alt+ctrl+shift+click (road)
alt+ctrl+click (junction)

**Comment by aubergine10** [comment]
@kvakvs @krzychu124 : Please test this PR. I'd ideally like to get it published to LABS this weekend so end-users can start testing. It's the main "new feature" of v11.1 release.

**Comment by kvakvs** [comment]
Motorway merge 2+1=3
Lane arithmetic is observed
Yield is produced too
bild

**Comment by aubergine10** [comment]
Good catch!!

**Comment by kianzarrin** [comment]
OK made 2 fixes:
Fix number 1: fix for counting lanes of Asym roads.
Fix number 2: better isSemiRoundabout detection and I call fix functions from roundabout utility.

Screanshot 1: Only Fix number 1.
ScreanShot 2 and 3: Both fixes.

Screenshot (426)
Screenshot (428)
Screenshot (429)

**Comment by krzychu124** [comment]
Testing...

**Comment by krzychu124** [comment]
@aubergine10 is it me, or delete button was changed from `recycle bin` to `x`

**Comment by kianzarrin** [comment]
> @krzychu124  I'm worried a bit about those `lock` thingies (I think it won't work - you probably need to lock whole instance).

Whole instance of what? Can you please elaborate. I don't have too much experience in multi-threading. What can go wrong here? what should I do?

@aubergine10 @kvakvs  any ideas?

**Comment by kianzarrin** [comment]
> 
> 
> @aubergine10 is it me, or delete button was changed from `recycle bin` to `x`

I has always been x for me. maybe I should merge.

**Comment by krzychu124** [comment]
`v10.21.1`
vlcsnap-error890

Regards to locking: https://stackoverflow.com/a/505604

**Comment by aubergine10** [comment]
IMO that button doesn't matter too much; I think at some point the Priority Signs UI workflow needs changing to something more like junction restrictions. So I wouldn't delay merging this becaue of that button.

**Comment by krzychu124** [comment]
Just noticed a change when I was testing priority bug, you can merge it :wink:

**Comment by kianzarrin** [comment]
> `v10.21.1`
> vlcsnap-error890
> Regards to locking: https://stackoverflow.com/a/505604

**Comment by kianzarrin** [comment]
> Regards to locking: https://stackoverflow.com/a/505604
@aubergine10 I don't see it. In the link you sent Its returning a class by ref. I do not return a class. I return bolean. `DateTime` is also a struct (which i don't return). I am not even using any byref values let alone mishandling them.

Am I missing something. is there another problem maybe?

**Pull request merged**